WIP: FEATURE: Rework CR CatchUp mechanism

bwaidelich opened the pull request:
related: #4746

WIP because:

- [ ] Proper(er) mechanism to prevent multiple catchup runs (maybe integrate #4751)?
- [ ] Mechanism to skip projections that are not up-to-date and don't trigger hooks on replay/individual catchup – (maybe integrate with #4992)
- [x] needs to be properly coordinated with other pending PRs like #4993, #4980, #4979
- [ ] update neos docs to not promote `block` see https://docs.neos.io/guide/contributing-to-neos/event-sourced-content-repository/blocking-until-a-command-is-executed and related

bwaidelich committed:
WIP: FEATURE: Rework CR CatchUp mechanism

bwaidelich committed:
Remove `react/async` dependency

as discussed in todays weekly

bwaidelich committed:
Remove `react/async` dependency 2/2

bwaidelich committed:
Cosmetic fix to satisfy linter

bwaidelich committed:
Remove unused namespace imports

bwaidelich committed:
Replace `CheckpointStorageInterface`

and introduce `neos/contentrepository-dbaltools` package

bwaidelich committed:
Replace `CheckpointStorageInterface`

and introduce `neos/contentrepository-dbaltools` package

bwaidelich committed:
Merge branch 'feature/4746-rework-catchup-mechanism' of https://github.com/neos/neos-development-collection into feature/4746-rework-catchup-mechanism

bwaidelich committed:
Remove `DbalClientInterface` and implementations

...and use `Doctrine\DBAL\Connection` directly

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/Tests/Behavior/Features/Bootstrap/FeatureContext.php:
```diff
@@ -44,7 +44,6 @@ public function __construct()
         self::bootstrapFlow();
         $this->contentRepositoryRegistry = $this->getObject(ContentRepositoryRegistry::class);
 
-        $this->setupCRTestSuiteTrait();
```
no longer needed, see https://github.com/neos/neos-development-collection/blob/9ee2fc85b30d03c2b5c176e2ec5409db555e74da/Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/CRTestSuiteTrait.php#L88

bwaidelich committed:
Update composer manifest

bwaidelich committed:
Tweak checkpoint error message

bwaidelich committed:
Re-add `neos/eventstore-doctrineadapter` dependency

bwaidelich committed:
Mostly cosmetic tweaks to satisfy linter

bwaidelich committed:
Make checkpoint table check a platform specific check

..to avoid status from showing a diff everytime

bwaidelich committed:
Remove notion of "sync" from composer scripts and readme

bwaidelich committed:
Remove ProjectionCatchUpTriggerInterface and implementations

...for now

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -153,6 +159,67 @@ public function projectionState(string $projectionStateClassName): ProjectionSta
+    {
+        $store = new SemaphoreStore();
+        $factory = new LockFactory($store);
+        $lock = $factory->createLock('catchup');
```
TODO: This lock needs to be CR specific (i.e. include `$this->id`)

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -153,6 +159,67 @@ public function projectionState(string $projectionStateClassName): ProjectionSta
 
+    public function catchUpProjections(): void
+    {
+        $store = new SemaphoreStore();
```
This has to become an external dependency and/or coordinated with #4751

nezaniel commented:
This already looks very promising and I like the direction this is taking!
What part is still missing / incomplete? Just the new catchup implementation, right?

bwaidelich commented:
> This already looks very promising and I like the direction this is taking!

Thanks for the feedback, appreciated!

> What part is still missing / incomplete?

I added some more lines to the task list above

mhsdesign commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Bootstrap/FeatureContext.php:
```diff
@@ -59,8 +59,7 @@ public function __construct()
 
-        $this->dbalClient = $this->getObject(DbalClientInterface::class);
-        $this->setupCRTestSuiteTrait();
+        $this->dbal = $this->getObject(Connection::class);
```
why do we use the `Connection` here directly instead of the outside world `DbalClientInterface` provider? ... in the end its the same ... must be i guess?

mhsdesign commented on Neos.ContentRepository.BehavioralTests/Tests/Functional/Feature/WorkspacePublication/setup-is-running-flag:
why do we need this file now? ^^

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -84,16 +86,12 @@ public function __construct(
-    public function handle(CommandInterface $command): CommandResult
+    public function handle(CommandInterface $command): object
     {
+        #print_r(debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS));exit;
+        #\Neos\Flow\var_dump('HANDLE ' . $command::class);
```
wip

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -127,7 +125,15 @@ public function handle(CommandInterface $command): CommandResult
 
-        return $this->eventPersister->publishEvents($eventsToPublish);
+        $this->eventPersister->publishEvents($this, $eventsToPublish);
+        return new class {
+            /**
+             * @deprecated backwards compatibility layer
+             */
+            public function block(): void
+            {
+            }
+        };
```
if we were to merge this like this - which i think is not necessary - we should use an actual class so its inspectable via phpstorm :) But i dont think we need b/c layer for new code

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -163,36 +230,26 @@ public function catchUpProjection(string $projectionClassName, CatchUpOptions $o
-            }
             $catchUpHook?->onBeforeEvent($event, $eventEnvelope);
             $projection->apply($event, $eventEnvelope);
+            // TODO this should happen in the inner transaction
```
todo over here :)

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -153,6 +159,67 @@ public function projectionState(string $projectionStateClassName): ProjectionSta
+            ];
+            $projectionsAndCatchUpHooks[$projectionClassName]['catchUpHook']?->onBeforeCatchUp();
+        }
+        #\Neos\Flow\var_dump('CATCHUP from ' . $lowestAppliedSequenceNumber->value);
```
.. lots of debug code and i guess todos?

mhsdesign commented on Neos.ContentRepository.Core/Classes/Infrastructure/DbalClientInterface.php:
what is the idea behind the removal of this?

mhsdesign commented on Neos.ContentRepository.Core/Classes/Infrastructure/DbalCheckpointStorage.php:
i guess this is now part of `CheckpointHelper`? Good thing we have it in the Core and now our lower level package ^^

mhsdesign reviewed (commented):
Thats a lot of stuff but looks "good" so far (considering this is WIP) ... i went through all php classes by reading except the last 20 with only the +-1 changes

bwaidelich committed:
Merge branch '9.0' into feature/4746-rework-catchup-mechanism

bwaidelich committed:
Merge branch '9.0' into feature/4746-rework-catchup-mechanism

bwaidelich committed:
Merge branch '9.0' into feature/4746-rework-catchup-mechanism

bwaidelich committed:
Fix ContentStreamProjection checkpoint

bwaidelich committed:
Merge branch '9.0' into feature/4746-rework-catchup-mechanism

bwaidelich committed:
Merge branch '9.0' into feature/4746-rework-catchup-mechanism

bwaidelich committed:
BUGFIX: Remove doctrine migrate and cr setup commands from behat scripts

See #5005

bwaidelich committed:
Use dedicated connection for EventStore

in order to prevent nested transactions potentially leading to RuntimeException of

    A transaction is active already, can't commit events!

bwaidelich committed:
Re-use previously created event store instances

bwaidelich committed:
Merge branch '9.0' into feature/4746-rework-catchup-mechanism

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/4746-rework-catchup-mechanism

mhsdesign committed:
TASK: Adjustments after merging 9.0 (beta10) to satisfy phpstan mostly

mhsdesign commented:
i just merged 9.0 into your pr. A few notes from my side:

- \ContentGraphTableNames::tableNamePrefix is now public again because its passed to the separate package `DbalTools` this makes `ContentGraphTableNames::checkpoint` obsolete as well ... should be adjusted
- i started merging the `.feature` behat tests but at last i decided to just accept 9.0 as it was getting tedious. My assumption based on the features i have seen changed is that only the blocking part was removed. As discussed ill split this up into a separate pr either way.
  - -> https://github.com/neos/neos-development-collection/pull/5059
- the (new - from the frankfurt sprint) shallow packages Neos.ContentRepositoryRegistry.DoctrineDbalClient Neos.ContentRepositoryRegistry.PostgresDbalClient seem to be now obsolete which is good imo (we have too much stuff first level for any new contributor. They should be removed.
- Due to only having the `CommandHandlingDependencies` at hand we cannot pass the cr to `publishEvents` to block.

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/4746-rework-catchup-mechanism

mhsdesign commented on Neos.ContentRepository.NodeMigration/src/Transformation/AddDimensionShineThroughTransformationFactory.php:
as discussed in https://github.com/neos/neos-development-collection/pull/5061 we want to keep the result thing for now and with it the return values here.

```
git checkout 9.0 -- Neos.ContentRepository.Core/Classes/Service/ContentStreamPruner.php
git checkout 9.0 -- Neos.ContentRepository.NodeMigration
```

bwaidelich committed:
Merge branch '9.0' into feature/4746-rework-catchup-mechanism

bwaidelich committed:
Merge branch 'task/remove-dbal-client' into feature/4746-rework-catchup-mechanism

# Conflicts:
#	Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php
#	Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/ProjectionIntegrityViolationDetector.php
#	Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ProjectionContentGraph.php
#	Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Projection/HypergraphProjection.php
#	Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Projection/ProjectionHypergraph.php
#	Neos.ContentRepository.Core/Classes/Projection/ContentStream/ContentStreamProjection.php
#	Neos.ContentRepository.Core/Classes/Projection/Workspace/WorkspaceProjection.php
#	Neos.ContentRepositoryRegistry.DoctrineDbalClient/Configuration/Objects.yaml
#	Neos.ContentRepositoryRegistry.DoctrineDbalClient/Configuration/Settings.yaml
#	Neos.ContentRepositoryRegistry.DoctrineDbalClient/composer.json
#	Neos.ContentRepositoryRegistry.PostgresDbalClient/Configuration/Objects.yaml
#	Neos.ContentRepositoryRegistry.PostgresDbalClient/composer.json
#	Neos.ContentRepositoryRegistry/Classes/Command/ContentGraphIntegrityCommandController.php
#	Neos.ContentRepositoryRegistry/Configuration/Objects.yaml
#	Neos.ContentRepositoryRegistry/Configuration/Settings.yaml
#	Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php

bwaidelich committed:
Merge branch '9.0' into feature/4746-rework-catchup-mechanism

bwaidelich committed:
Make ContentGraphTableNames::tableNamePrefix private again

mhsdesign commented on phpstan.neon.dist:
```diff
@@ -9,13 +9,17 @@ parameters:
         - Neos.ContentRepository.StructureAdjustment/src
         - Neos.ContentRepository.TestSuite/Classes
         - Neos.ContentRepositoryRegistry/Classes
+        - Neos.ContentRepositoryRegistry.DoctrineDbalClient
+        - Neos.ContentRepositoryRegistry.PostgresDbalClient
```
```suggestion
```

they have been removed again, my change can thus be reverted

bwaidelich replied:
I'll take care of the rebasing once my three cleanups #5073 #5074 #5075 are in

bwaidelich committed:
Merge branch '9.0' into feature/4746-rework-catchup-mechanism

bwaidelich commented:
Closing because most of this is obsolete and will be fixed with #5321
